eBPF processing

niclashedam opened the pull request:
I would like a preliminary review on the progess I've made so far on this task. I know there is a lot of details missing, but I would like to know if the overall structure looks reasonable.

@iomartin, do you have time to look at it?

iomartin commented on hw/misc/hermes.c:
```diff
@@ -84,6 +84,15 @@ static const struct hermes_bar0 {
     uint32_t ehdsoff;
     uint32_t ehdssze;
+
+    char padding1[0x1000 - 0x48]; /* Padding between 0x48 and 0x1000 */
```
I'd say we don't need to have the paddings in the struct. hermes_bar0 doesn't have to match the spec bit for bit, we can use logic in the read/write functions to get the right struct member rather than dealing just with offsets.

niclashedam replied:
Fair, this is good news. I got a bit tired of calculating offsets.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -84,6 +84,15 @@ static const struct hermes_bar0 {
+
+    char padding1[0x1000 - 0x48]; /* Padding between 0x48 and 0x1000 */
+
+    char ehcmdreq0[32];
```
I recommend defining structs with the command request/response fields rather than making them char arrays. It will make the code a lot more readable and avoid having to deal with hardcoded offsets all over

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1152,3 +1182,4 @@ static void pci_hermes_register_types(void)
     type_register_static(&hermes_info);
 }
 type_init(pci_hermes_register_types)
+
```
extra empty line

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1070,6 +1072,14 @@ static void hermes_instance_init(Object *obj)
     }
+
+    for (i = 0; i < UBPF_ENGINES; i++) {
+        hermes->engines[i] = ubpf_create();
```
This should probably be on the pci_hermes_realize() function. A device is constructed in two steps: init and realize. This is the [difference](https://github.com/qemu/qemu/blob/48202c712412c803ddb56365c7bca322aa4e7506/include/hw/qdev-core.h#L48) between them:

    * # Realization #
    * Devices are constructed in two stages,
    * 1) object instantiation via object_initialize() and
    * 2) device realization via #DeviceState:realized property.
    * The former may not fail (and must not abort or exit, since it is called
    * during device introspection already), and the latter may return error
    * information to the caller and must be re-entrant.
    * Trivial field initializations should go into #TypeInfo.instance_init.
    * Operations depending on @props static properties should go into @realize.
    * After successful realization, setting static properties will fail.

Because ubpf_create() can fail, we should do it in the realize and propagate the error up if it fails.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -88,14 +88,10 @@ static struct hermes_bar0 {
-    char padding2[0x1000 - 48]; /* Padding between 0x1030 and 0x2000 */
-
-    uint8_t ehcmdctrl0;
+    char commands1[0x1000];
```
This is wasting space, only allocate memory for the number of UBPF_ENGINES we have

iomartin commented on hw/misc/hermes.c:
```diff
@@ -435,12 +429,30 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
-    default:
+    }
+
+    else if (addr >= 0x2000) {
```
the else if should be in the same line as the closing bracket

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1075,6 +1088,7 @@ static void hermes_instance_init(Object *obj)
 
     for (i = 0; i < UBPF_ENGINES; i++) {
         hermes->engines[i] = ubpf_create();
+        memcpy(&bar0_init + 0x2000 + i, &zero, 1);
```
You can use memset here (and it can be out of the loop)

iomartin replied:
Also, instead of doing a `bar0_init + 0x2000`, you could do `bar0_init.commands2`

iomartin replied:
Also, instead of doing a `bar0_init + 0x2000`, doing `bar0_init.commands2` is more legible and won't introduce bugs if we change offsets later

iomartin replied:
Also, instead of doing a `bar0_init + 0x2000`, doing `bar0_init.commands2` is more legible and won't introduce bugs if we change offsets later

iomartin replied:
And I'd probably memset the command request/responses too

iomartin commented on hw/misc/hermes.c:
```diff
@@ -435,12 +429,30 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
+        int engine = (addr - 0x1000) / 48;
+        int offset = (addr - 0x1000) % 48;
+
+        if (offset != 0 || engine > UBPF_ENGINES - 1) {
```
Why is `offset !=0` invalid? For example 0x1004 is fine.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -88,14 +88,10 @@ static struct hermes_bar0 {
-
-    uint8_t ehcmdctrl0;
+    char commands1[0x1000];
+    char commands2[UBPF_ENGINES];
```
I'd probably call this somethign like cmds_ctrl

iomartin commented on hw/misc/hermes.c:
```diff
@@ -449,7 +472,11 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
 
-        memcpy(ptr, &val, size);
+        if (val == 1) {
+            memcpy(ptr, &val, 1);
```
Use memset

iomartin commented on hw/misc/hermes.c:
```diff
@@ -393,6 +398,23 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+
+    if (command != 0x80) {
+        hermes_bar_warn_invalid_command(command, engine);
+        return 0x1; /* TODO: Is there an error code for invalid command? */
```
Currently there isn't an error code for invalid command. Can you please open a PR against https://github.com/Eideticom/eid-hermes/blob/master/specs/eid-hermes-commands-format.md?

niclashedam replied:
https://github.com/Eideticom/eid-hermes/pull/31

iomartin commented on hw/misc/hermes.c:
```diff
@@ -398,21 +399,52 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+    void *data_base = bar4_base + bar0_init.ehdsoff +
+                      (data_slot * bar0_init.ehdssze);
+
+    char *errmsg;
```
nit: all variables should be defined in the beginning of the function

iomartin commented on hw/misc/hermes.c:
```diff
@@ -398,21 +399,52 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+    bool elf;
+
+    /* Check magic number (first 4 bytes), to see if program is in ELF format */
+    elf = bar0_init.ehpssze >= 4 && !memcmp(program_base, ELFMAG, 4);
```
Instead of the hardcoded 4, I'd use `SELFMAG`, just like they do in ubpf

iomartin commented on hw/misc/hermes.c:
```diff
@@ -398,21 +399,52 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+    }
+
+    if (rv < 0) {
+        response[32] = 0x05;
```
0x05 should be a `#define` or part of an enum with all other possible status codes

iomartin commented on hw/misc/hermes.c:
```diff
@@ -30,6 +30,7 @@
 #include "qapi/error.h"
 #include "trace.h"
 #include <ubpf.h>
+#include <elf.h>
```
What's MVP?

niclashedam replied:
_Minimum viable product_
I use the term to state that we are at the stage, where the product (in this case PR) are around the point where we could potentially integrate this, altough it would be far from perfect.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -393,6 +398,23 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+    void *base = (uint32_t *) &((uint8_t *) &bar0_init)[0x1000 + engine * 48];
+    int command = *(int *) base;
+
+    if (command != 0x80) {
```
0x80 should be a `#define` or part of an enum with all other possible opcodes

iomartin replied:
0x80 should be a `#define` or part of an enum with all other possible opcodes

iomartin commented on hw/misc/hermes.c:
```diff
@@ -393,6 +398,23 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
                 addr);
 }
 
+static uint64_t hermes_exec(void *opaque, int engine)
```
Instead of a `void *opaque`, I'd make this a `HermesState *hermes`

iomartin reviewed (commented):
Thanks @niclashedam. I have added some comments.

However, I found this PR could have been easier to follow. First, I'd like to see commits with more than a title. Also, I feel like many commits could have been squashed to tell a clearer story. In a few instances, a commit would add some code and the subsequent commit would replace it with something else and often we only need to see the final version. Or they added only part of a functionality. For a reviewer, it is much easier when a commit is concise and self-contained ;)

iomartin commented on hw/misc/hermes.c:
```diff
@@ -383,6 +399,54 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+        return; /* TODO: Is there an error code for invalid command? */
+    }
+
+    int program_slot = *(int *)(base + 8);
```
You should also check that the program slot and data slot are valid

niclashedam commented:
I've integrated the suggested changes. What about now?

I'm sorry about the mess. I've squashed all comments into a single one.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -383,6 +423,64 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+    unsigned char prog_slot = command->req_meta[0];
+    unsigned char data_slot = command->req_meta[1];
+
+    if (prog_slot > bar0_init.ehpslot) {
```
this should be >= (and same thing below)

iomartin commented on hw/misc/hermes.c:
```diff
@@ -69,9 +73,35 @@
-static const struct hermes_bar0 {
+struct hermes_bar0_command {
+    /* Request */
+    unsigned char req_opcode;
```
We should use fixed size types for these (uint8_t, etc)

iomartin commented on hw/misc/hermes.c:
```diff
@@ -69,9 +73,35 @@
+    unsigned char reserved0;
+    unsigned short req_identifier;
+    unsigned int reserved1;
+    unsigned char req_meta[24];
```
I'd suggest you make this into a union, so that we can access the command-specific fields more easily. Something like [this](https://github.com/Eideticom/eid-hermes/blob/402ddd10b9ad82dc680ea51aa1bc8b7aafdca843/src/driver/hermes_mod.h#L91).

iomartin commented on hw/misc/hermes.c:
```diff
@@ -69,9 +73,35 @@
+    unsigned char res_meta[8];
+};
+
+static struct hermes_bar0 {
```
Now that hermes_bar0 is no longer `const`, we should probably make all the old fields `const` so that we don't accidentally overwrite any of them.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -383,6 +423,64 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+        return; /* TODO: Is there an error code for invalid command? */
+    }
+
+    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
```
All variable declarations should be in the beginning of the function

iomartin commented on hw/misc/hermes.c:
```diff
@@ -383,6 +423,64 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+
+    if (rv < 0) {
+        command->res_status = HERMES_STATUS_EBPF_ERROR;
+        command->res_meta[0] = rv;
```
This is downcasting the 32 bytes `rv` to the 8 bytes `res_meta[0]`

iomartin commented on hw/misc/hermes.c:
```diff
@@ -84,10 +114,13 @@ static const struct hermes_bar0 {
     uint32_t ehdssze;
+
+    struct hermes_bar0_command commands[UBPF_ENGINES];
+    char exec[UBPF_ENGINES];
```
As discussed in the [driver PR](https://github.com/Eideticom/eid-hermes/pull/23#discussion_r561433069), we'll need a way of telling the host that execution has finished. My suggestion is to use the 0x2000 registers (EHCMDCTRL) for that. The driver could write to bit 0 to start program execution, then the device writes to bit 1 to indicate that the program has finished.

Thus (and to keep this consistent with the spec), I'd call this `cmdctrl` instead of `exec`.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -399,18 +497,61 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
             val &= 0xFFFF;
             break;
         }
+    } else if (addr >= 0x1000 &&
```
This is fine I think, but we need to fix the `if` above. `sizeof(struct hermes_bar0)` has increased, so it's no longer appropriate to use it in the check

iomartin commented on hw/misc/hermes.c:
```diff
@@ -399,18 +497,61 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
+               addr + size <= 0x1000 +
+               sizeof(struct hermes_bar0_command) * UBPF_ENGINES) {
+        ptr = (uint32_t *) &((uint8_t *) &bar0_init.commands)[addr];
+        val = *ptr;
```
If `size != 4`, then we would be reading more than requested. I think the easiest solution is just to move the `switch (size)` to right before the return.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -399,18 +497,61 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
                 unsigned size)
 {
-    hermes_bar_warn_read_only(0, addr);
+    if (addr + size > sizeof(struct hermes_bar0)) {
```
same comment as above about `sizeof(struct hermes_bar0)`

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1244,15 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+
+    for (i = 0; i < UBPF_ENGINES; i++) {
+        hermes->engines[i] = ubpf_create();
+        memcpy(&bar0_init + 0x2000 + i, &zero, 1);
```
Use memset. And do this outside of the loop.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -399,18 +497,61 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
+            return;
+        }
+
+        if (val == 1) {
```
As discussed above, we should look at bit 0 only.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -399,18 +497,61 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
+
+        if (val == 1) {
+            bar0_init.exec[engine] = 1;
+            hermes_exec(hermes, engine);
```
The eBPF program should be executed into a different thread, otherwise we block the main qemu. Take a look at the XDMA code for an example. Good starting points are `hermes_bar2_init_dir_reg` and `hermes_dma_thread`.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1244,15 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+        memcpy(&bar0_init + 0x2000 + i, &zero, 1);
+
+        if (!hermes->engines[i]) {
+            error_report("Failed to create uBPF engine %i", i);
```
I don't think we want to just report that it failed, we probably want QEMU to abort. We can do that with error_propagate().

iomartin reviewed (commented):
See comments. I'd still like to see a more descriptive commit message (and move the resolves tag to the commit message).

Have you tried running this with https://github.com/Eideticom/eid-hermes/pull/23? While that PR is still missing the interrupt handling (it expects the eBPF execution to be "instantaneous"), it has most of the work done (I think), writing all the expected bits on the expected offsets.

We want to keep those two PRs in sync, so I recommend you follow that PR.

niclashedam commented:
Thank you 😄 I've fixed all of your comments, but two.

First, I would like to have a more clear idea on how you want the eBPF thread to work. Should I just spawn a new thread using `qemu_thread_create` ? Should I set up mutexes or is it up the the user to not use the same engine at the same time?

In regards to the union; Can you explain how you want it to work? I have not worked with unions before.

By the way; I'm sorry that things move a little slow currently. I'm still trying to get used to work from home.

iomartin commented:
Thanks @niclashedam.

> First, I would like to have a more clear idea on how you want the eBPF thread to work. Should I just spawn a new thread using qemu_thread_create ? Should I set up mutexes or is it up the the user to not use the same engine at the same time?

Yes, we should spawn a new `QemuThread` for each eBPF engine, with a `QemuMutex` and `QemuCond` for each as well. While the kernel driver will not submit two jobs to an eBPF engines at the same time, we should design our code to be robust.

See for example the hermes_dma_thread function:

```C
static void hermes_dma_thread(HermesState *hermes,
                              struct hermes_bar2_dir_reg *dir)
{
    qemu_mutex_lock(&dir->dma_mutex);
    while (1) {
        qemu_cond_wait(&dir->dma_cond, &dir->dma_mutex);
        if (hermes->stopping) {
            break;
        }
        do_dma(hermes, dir);
    }
    qemu_mutex_unlock(&dir->dma_mutex);
}
```

This function blocks on `qemu_cond_wait()` until `dma_cond` is signaled, which happens on `hermes_bar2_engine_write`:

```C
static uint64_t hermes_bar2_engine_write(HermesState *hermes,
                                         struct hermes_bar2_dir_reg *dir,
                                         hwaddr addr, uint32_t val)
{
    struct hermes_bar2_engine_reg *reg = &dir->channel;
    hwaddr masked_addr = addr & 0xFF;

    switch (masked_addr) {
    case 0x04:
        reg->control = 0x0FFFFE7F & val;
        if (reg->control & 0x1) {
            qemu_mutex_lock(&dir->dma_mutex);
            qemu_cond_signal(&dir->dma_cond);
            qemu_mutex_unlock(&dir->dma_mutex);
        }
        break;
...
```

See the manpage of [`pthread_cond_wait`](https://linux.die.net/man/3/pthread_cond_wait) and [`pthread_cond_signal`](https://linux.die.net/man/3/pthread_cond_signal) for more details about these functions (the qemu versions are just wrappers around the pthread ones)

> In regards to the union; Can you explain how you want it to work? I have not worked with unions before.

In C, unions resemble structs, as they are a grouping of multiple members, but with the important distinction that, in unions, all members use the same memory location, while in structs each member has its own memory location. Unions are very useful for when you want to interpret the same memory location in different ways. A union's size is the size of the largest member. See [here](https://www.tutorialspoint.com/cprogramming/c_unions.htm) for more details.

You could do something like this (this is taken form the the [kernel PR](https://github.com/Eideticom/eid-hermes/blob/3fd3bb0d32dcd4b587feef12d074fd58f5e6d216/src/driver/hermes_mod.h#L91) (https://github.com/Eideticom/eid-hermes/pull/23)):

```C
struct __attribute__((__packed__)) hermes_cmd_req {
	uint8_t opcode;
	uint8_t rsv0;
	uint16_t cid;
	uint32_t rsv1;
	union {
		struct __attribute__((__packed__)) {
			uint8_t prog_slot;
			uint8_t data_slot;
		} run_prog;

		uint8_t cmd_specific[24];
	};
};

struct __attribute__((__packed__)) hermes_cmd_res {
	uint16_t cid;
	uint8_t status;
	uint8_t rsv0[5];
	union {
		struct __attribute__((__packed__)) {
			uint64_t ebpf_ret;
		} run_prog;

		uint8_t cmd_specific[8];
	};
};

struct __attribute__((__packed__)) hermes_cmd {
	struct hermes_cmd_req req;
	struct hermes_cmd_res res;
};
```

iomartin commented:
Hey @niclashedam, one other thing that we'll need to address are interrupts. After the driver submits an eBPF job, it needs a way to know when execution has finished.

I think we can do something similar for what XDMA does. You'll probably need to assign an irq_vector for each eBPF engine, and set them to incremental values starting at `H2C_CHANNELS + C2H_CHANNELS` (the first irq numbers are used for XDMA, see `hermes_instance_init()`). Then when an eBPF program finishes, we'd call `msix_notify()`. Also see https://github.com/Eideticom/eid-hermes/commit/65a6a04e1d0b7ec629f7900b0cfe5e31a19f3344 for a recent change to the BAR0 spec.

I have just pushed a newer version of the driver code that starts handling the interrupts. The code is still incomplete (the interrupt handler doesn't do anything yet), but it can serve as a starting point.

niclashedam commented:
Hey. I added the thread, unions and interrupt. Everything seems OK but a compile error that I cannot get my head around. Let me know what you think.

```
make[1]: Entering directory '/home/nhed/Development/eid-hermes-qemu/slirp'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/nhed/Development/eid-hermes-qemu/slirp'
  CC      hw/misc/hermes.o
hw/misc/hermes.c: In function ‘pci_hermes_realize’:
hw/misc/hermes.c:1315:31: error: assignment to ‘void * (*)(void *)’ from incompatible pointer type ‘void (*)(void *)’ [-Werror=incompatible-pointer-types]
 1315 |         hermes->ubpf[i].entry = &hermes_bpf_thread;
      |                               ^
cc1: all warnings being treated as errors
make: *** [/home/nhed/Development/eid-hermes-qemu/rules.mak:69: hw/misc/hermes.o] Error 1
```

iomartin commented:
I haven't reviewed the code, but the error you're getting is because `hermes_bpf_thread` is of wrong type. It should be `static void *hermes_bpf_thread(void *opaque)`.

niclashedam commented:
Ah, thanks. Sometimes the error messages in C and C++ are a bit hard to decipher.

Commit:
Setup uBPF engines

This commit implements the instantiation and tear down of uBPF engines.

Signed-off-by: Niclas Hedam <nhed@itu.dk>

Commit:
Enable uBPF program execution

This commit allows the triggering and execution of uBPF programs.

Resolves #8.

Signed-off-by: Niclas Hedam <nhed@itu.dk>

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1148,30 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+            QEMU_THREAD_JOINABLE);
+
+        if (!hermes->ubpf[i].engine) {
+            error_propagate(errp, NULL);
```
The second argument can't be NULL, doing that effectively bypasses error_propagate (and error_propagate should terminate the program)

You will need to create a new `Error`, set it with `error_setg()`and pass it to error_propagate.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1148,30 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+            QEMU_THREAD_JOINABLE);
+
+        if (!hermes->ubpf[i].engine) {
+            error_propagate(errp, NULL);
```
The second argument can't be NULL, doing that effectively bypasses error_propagate (and error_propagate should terminate the program)

You will need to create a new `Error`, set it with `error_setg()` and pass it to error_propagate.

iomartin replied:
And you can return from this function right after we call error_propagate, there is no point in continuing.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1148,30 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+            &hermes->ubpf[i],
+            QEMU_THREAD_JOINABLE);
+
+        if (!hermes->ubpf[i].engine) {
```
This should be right after the call to ubpf_create(). If ubpf_create() fails, we don't need to do the rest

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1148,30 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+    }
+
+    memset(&bar0_init.cmdctrl, 0, UBPF_ENGINES
+                                  * sizeof(struct hermes_cmd_ctrl));
```
It doesn't hurt to also init bar0_init.commands

niclashedam replied:
Since they are neighbours I could do it in one run, but I choose to do it in two for safety.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -159,6 +192,17 @@ struct hermes_bar2 {
+    QemuThread bpf_thread;
+
+    int no;
+    void *(*entry)(void *);
```
We don't need the entry field in the struct, all threads use the same entry function, so we can use it directly when calling qemu_thread_create()

iomartin commented on hw/misc/hermes.c:
```diff
@@ -159,6 +192,17 @@ struct hermes_bar2 {
     struct hermes_bar2_cfg_reg cfg;
 };
 
+struct hermes_ubpf_reg {
```
This is not a register, let's call it `hermes_ubpf_eng`

iomartin commented on hw/misc/hermes.c:
```diff
@@ -95,6 +95,48 @@
 
 typedef struct HermesState HermesState;
 
+struct __attribute__((__packed__)) hermes_cmd_req
```
Can we simplify this struct what is used on the driver code?

https://github.com/Eideticom/eid-hermes/blob/948b8f00a6e391807b67a39c0f1d1a0c5288d6c7/src/driver/hermes_mod.h#L90

iomartin replied:
And also hermes_cmd_res

niclashedam replied:
You want me to copy and paste the structs defined in the driver?

iomartin replied:
Yes

iomartin commented on hw/misc/hermes.c:
```diff
@@ -445,18 +487,65 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
 {
-    hermes_bar_warn_read_only(0, addr);
+    if (addr >= HERMES_BAR0_CMD_REQ && addr < HERMES_BAR0_CMD_CTRL) {
+        int engine = (addr - HERMES_BAR0_CMD_REQ) / 48;
```
nit: variables should be declared in the beginning of the function.

iomartin replied:
Instead of hardcoding 48, you can use `sizeof(struct hermes_cmd)`

iomartin commented on hw/misc/hermes.c:
```diff
@@ -69,25 +73,54 @@
+#define HERMES_STATUS_EBPF_ERROR        0x05
+#define HERMES_STATUS_INVALID_OPCODE    0x06
+
+#define HERMES_CMD_STOP           0x0
```
nit: Similarly, `HERMES_CMD` defines are only used in the second commit, so add them there

iomartin commented on hw/misc/hermes.c:
```diff
@@ -69,25 +73,54 @@
 
+#define HERMES_OPCODE_REQUEST_SLOT 0x00
+#define HERMES_OPCODE_RELEASE_SLOT 0x01
+#define HERMES_OPCODE_EXECUTE_SLOT 0x80
```
nit: the HERMES_OPCODE_* and HERMES_STATUS_* defines are only used in the last commit of this series, add them there.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -445,18 +487,65 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
+        }
+    }
+
+    else {
```
nit: this should be an else-if/else. And it should be on the same line as the closing bracket above.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -459,6 +459,11 @@ static inline void hermes_bar_warn_invalid(unsigned bar, hwaddr addr)
     warn_report("Hermes: Accessed invalid BAR%d register: 0x%lX", bar, addr);
 }
 
+static inline void hermes_bar_warn_invalid_command(int command, int engine)
```
Nit: add the "resolves #8" tag on its own line, as the last line of the commit message.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -471,6 +476,97 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+        return;
+    }
+
+    command->res.cid = command->req.cid;
```
This should be above the prog_slot and data_slot checks. Even if the slots are invalid, we want to set the response cid

iomartin commented on hw/misc/hermes.c:
```diff
@@ -471,6 +476,97 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+
+    if (prog_slot >= bar0_init.ehpslot) {
+        command->res.status = HERMES_STATUS_INVALID_PROG_SLOT;
+        mark_exec_finished(hermes, engine);
```
We're repeating the pattern

```
mark_exec_finished(hermes, engine);
return;
```

in many places of the code, including in the end. We can simplify this with a `goto` statement.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -471,6 +476,97 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+    }
+
+    if (rv < 0) {
+        command->res.status = HERMES_STATUS_EBPF_ERROR;
```
If the ubpf_load() fails, it writes an error message on `errmsg`. We should print this message with `warn_report()`

iomartin commented on hw/misc/hermes.c:
```diff
@@ -471,6 +476,97 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+
+    ret = ubpf_exec(hermes->ubpf[engine].engine, data_base, bar0_init.ehdssze);
+
+    command->res.status = HERMES_STATUS_SUCCESS;
```
ubpf_exec can fail (it returns UINT64_MAX), in which case we should set the status to HERMES_STATUS_EBPF_ERROR

iomartin commented on hw/misc/hermes.c:
```diff
@@ -383,12 +476,103 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+    /* Check magic number (first 4 bytes), to see if program is in ELF format */
+    elf = bar0_init.ehpssze >= SELFMAG && !memcmp(program_base, ELFMAG, 4);
+    if (elf) {
+        rv = ubpf_load_elf(hermes->ubpf[engine].engine, program_base,
```
We're going to have problems with this... The current ubpf_load* functions do not support loading a second program on top of the previous one.

I'm going to submit a PR to the ubpf project to address this, but in the meantime, can we add the following before trying to load?

```C
if (hermes->ubpf[engine].engine->insts) {
    free(hermes->ubpf[engine].engine->insts);
    hermes->ubpf[engine].engine->insts = NULL;
}
```

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1130,37 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+        hermes->ubpf[i].engine = ubpf_create();
+
+        if (!hermes->ubpf[i].engine) {
+            Error *err = NULL;
```
Variable declarations should be in the beginning of the function

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1130,37 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+
+        if (!hermes->ubpf[i].engine) {
+            Error *err = NULL;
+            error_setg(errp, "uBPF engine not setup correctly.");
```
Can we change this to "error creating uBPF engine"?

iomartin replied:
Oh, and you should not be setting errp, you should set err

iomartin commented on hw/misc/hermes.c:
```diff
@@ -427,18 +474,62 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
+
+        memcpy((void *) &bar0_init.commands[engine] + offset, &val, size);
+    } else if (addr >= HERMES_BAR0_CMD_CTRL) {
+        engine = addr - HERMES_BAR0_CMD_CTRL;
```
This should be

    engine = (addr - HERMES_BAR0_CMD_CTRL)/sizeof(struct hermes_cmd_ctrl);

iomartin commented on hw/misc/hermes.c:
```diff
@@ -458,6 +475,108 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+
+    if (prog_slot >= bar0_init.ehpslot) {
+        command->res.status = HERMES_STATUS_INVALID_PROG_SLOT;
+        goto FINISH_EXEC;
```
nit: goto labels are usually lowercase

iomartin commented on hw/misc/hermes.c:
```diff
@@ -458,6 +475,108 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+     *  https://github.com/iovisor/ubpf/pull/63
+     *  https://github.com/Eideticom/eid-hermes-qemu/pull/14#discussion_r622587139
+     */
+    if (hermes->ubpf[engine].engine->insts) {
```
Actually, I just realized that this isn't going to compile, `ubpf_vm` is an opaque type... Let's see if we can get https://github.com/iovisor/ubpf/pull/63 in so we can call ubpf_unload(). Otherwise we will need to create a new VM every time...

niclashedam replied:
I'll leave this be. iovisor/ubpf#63 is progressing, so we can do this change as the last thing before merging.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -458,6 +475,108 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+            command->res.status = HERMES_STATUS_EBPF_ERROR;
+    }
+
+    command->res.status = HERMES_STATUS_SUCCESS;
```
this should be in an else, otherwise we overwrite the status you just set above in the error case

niclashedam commented:
I have fixed all comments, but the unload one. We can fix this when we know how iovisor/ubpf#63 will look like. For the same reason the commits are not signed. The final ones will be signed.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -69,25 +73,37 @@
 #define W1C(old, new) ((old) & ~(new))
 
+#define HERMES_BAR0_CMD_REQ    0x1000
+#define HERMES_BAR0_CMD_CTRL   0x2000
```
These two defines aren't used until next commit, so they can be included there

iomartin commented on hw/misc/hermes.c:
```diff
@@ -69,25 +73,37 @@
-static const struct hermes_bar0 {
-    uint32_t ehver;
-    char ehbld[48];
+struct __attribute__((__packed__)) hermes_cmd_ctrl {
```
Similarly, the hermes_cmd and hermes_cmd_ctrl structs and fields are not really used in this commit, so they should be moved to the next one (make sure to move their declaration and initialization as well).

iomartin commented on hw/misc/hermes.c:
```diff
@@ -427,18 +460,62 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
+            return;
+        }
+
+        if ((val & HERMES_CMD_START) == 1) {
```
With the new spec that uses bytes instead of bits for the command control, this can now be written simply as

    if (val == 1)

iomartin replied:
Or actually `val == HERMES_CMD_START`

iomartin commented on hw/misc/hermes.c:
```diff
@@ -444,6 +461,108 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+    uint64_t ret;
+    bool elf;
+
+    unsigned char prog_slot = command->req.run_prog.prog_slot;
```
`run_prog` doesn't exist anymore. And there are a couple other compilation errors as well

iomartin commented on hw/misc/hermes.c:
```diff
@@ -444,6 +461,108 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+        command->res.status = HERMES_STATUS_EBPF_ERROR;
+    } else {
+        command->res.status = HERMES_STATUS_SUCCESS;
+        command->res.run_prog.ebpf_ret = ret;
```
`ebpf_ret` should be set in both cases (see the spec)

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1130,37 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+
+        qemu_thread_create(&hermes->ubpf[i].bpf_thread,
+            hermes->ubpf[i].name,
+            &hermes_bpf_thread,
```
`hermes_bpf_thread` doesn't exist yet, the thread creation (as well as the mutex and cond inits) should be moved to the last commit.

iomartin reviewed (changes requested):
Thanks, I have added some more comments.

niclashedam commented:
Fixed. The code should compile now too.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1212,15 +1324,23 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
         if (!hermes->ubpf[i].engine) {
-            Error *err = NULL;
-            error_setg(errp, "uBPF engine not setup correctly.");
+            error_setg(&err, "error creating uBPF engine");
```
this should be fixed in the first commit

iomartin commented on hw/misc/hermes.c:
```diff
@@ -415,18 +460,62 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
+
+        if ((val & HERMES_CMD_START) == 1) {
+            bar0_init.cmdctrl[engine].ehcmddone = HERMES_CMD_NOT_FINISHED;
+            qemu_mutex_lock(&hermes->ubpf[engine].bpf_mutex);
```
Hmm I just realized we can't lock and send the signal yet, those haven't been initialized yet. Let's move all of this if to the next commit.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -504,7 +615,7 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
         }
 
-        if ((val & HERMES_CMD_START) == 1) {
+        if (val == HERMES_CMD_START) {
```
this should have been fixed in the second commit

niclashedam replied:
This is not applicable any more, as this block of code was moved to the third commit.

iomartin reviewed (commented):
Getting closer now. Just need to get ubpf_unload_code() approved into ubpf

niclashedam commented:
Fixed the commits. I have also added the unload function under the assumption that it will become available very soon.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -415,18 +460,55 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
+            hermes_bar_warn_invalid(0, addr);
+            return;
+        }
+
```
nit: remove extra empty line

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1119,20 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
     }
+
+    for (i = 0; i < UBPF_ENGINES; i++) {
+        hermes->ubpf[i].engine = ubpf_create();
```
I just realized that we are missing calls to ubpf_destroy() - as the commit message says it does. These should be called on pci_hermes_exit().

niclashedam replied:
pci_hermes_exit ? Do you mean pci_hermes_uninit?

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1102,6 +1119,20 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+
+    for (i = 0; i < UBPF_ENGINES; i++) {
+        hermes->ubpf[i].engine = ubpf_create();
+
```
nit: remove remove empty line between ubpf_create() and the if below

iomartin commented on hw/misc/hermes.c:
```diff
@@ -1133,6 +1215,12 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
+
+    memset(&bar0_init.cmdctrl, 0, UBPF_ENGINES
+                                  * sizeof(struct hermes_cmd_ctrl));
+
```
nit: remove extra empty line

iomartin commented on hw/misc/hermes.c:
```diff
@@ -444,6 +461,103 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
+    ubpf_unload_code(hermes->ubpf[engine].engine);
+
+    /* Check magic number (first 4 bytes), to see if program is in ELF format */
+    elf = bar0_init.ehpssze >= SELFMAG && !memcmp(program_base, ELFMAG, 4);
```
nit: can you replace the '4' with 'SELFMAG'? This how ubpf vm/test.c does it

iomartin commented on hw/misc/hermes.c:
```diff
@@ -504,6 +618,13 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
+            qemu_mutex_lock(&hermes->ubpf[engine].bpf_mutex);
+            qemu_cond_signal(&hermes->ubpf[engine].bpf_cond);
+            qemu_mutex_unlock(&hermes->ubpf[engine].bpf_mutex);
+            bar0_init.cmdctrl[engine].ehcmdexec = HERMES_CMD_STOP;
```
That should be set only when we finish the command (i.e., on `mark_exec_finished()`). Remember that the program is actually executed on a different thread, we don't want to set this register too early ;)

niclashedam replied:
In mark_exec_finished it is set to FINISHED. Maybe we can reduce it to just START and STOP?

iomartin replied:
That would not be enough, as we can't detect a program in a running state.

iomartin reviewed (changes requested):
A few more nits and a couple of other fixes and then I think it'll be good to get merged (assuming we can get the ubpf new call merged as well)

iomartin commented on hw/misc/hermes.c:
```diff
@@ -399,18 +575,62 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
+
+        memcpy((void *) &bar0_init.commands[engine] + offset, &val, size);
+    } else if (addr >= HERMES_BAR0_CMD_CTRL) {
+        engine = (addr - HERMES_BAR0_CMD_CTRL) / sizeof(struct hermes_cmd_ctrl);
```
I just realized we have to check the offset here as well (not just the engine). Offset 0 is EHCMDEXEC and is RW, offset 1 is EHCMDDONE and is RO and offsets 2-7 are invalid.

niclashedam commented:
Checked the offset and moved the CMD_STOP to `mark_exec_finished()`.

Commit:
Update uBPF to newest version to enable unloading of programs.

niclashedam commented:
I updated the commit hash of uBPF to expose the functionality introduced in https://github.com/iovisor/ubpf/pull/63.
However, the building of QEMU fails as it cannot find uBPF. Can you check why this happens?

iomartin commented:
It's a problem with uBPF, I opened a PR there to fix it: https://github.com/iovisor/ubpf/pull/91

iomartin commented:
Also, when that gets merged, can you change the order of your commits so that the commit that changes the UBPF SHA is the first in this series?

iomartin commented:
They've already merged it :)

niclashedam commented:
It seems like uBPF made a breaking change, dividing the eBPF return value from the uBPF VM return value. I've refactored the code such that it work as before their change.

niclashedam commented:
What is the status on this @iomartin - can we get it merged?

Commit:
Bump version to 1 and increase offsets to 64-bit. See https://github.com/Eideticom/eid-hermes/pull/40.

iomartin commented on hw/misc/hermes.c:
```diff
@@ -404,7 +449,7 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
     uint32_t val = 0;
 
-    if (addr + size <= sizeof(struct hermes_bar0)) {
+    if (addr + size <= 0x48) {
```
In the last commit, when you increase the offsets to 8 bytes, this causes these addresses to change as well. It should be <= 0x50 now, Or, even better, let the compiler figure it out with

```c
if (addr + size <= offsetof(struct hermes_bar0, commands))
```

You'll also need to do this for the write path. Please do it in the same commit as when you change the offsets to 8 bytes

iomartin commented:
Oh and can you please rebase on master now that #17 was merged?

niclashedam commented:
I think this is ready to merge.
